p2p/src/sync/peer.rs

```diff
@@ -369,6 +383,38 @@ where
         Ok(())
     }
 
+    async fn handle_transaction_request(&mut self, id: Id<Transaction>) -> Result<()> {
+        if !self.services.has_service(Service::Transactions) {
+            return Err(P2pError::ProtocolError(ProtocolError::UnexpectedMessage(
+                "A transaction request is received, but this node doesn't have the corresponding service",
+            )));
+        }
+
+        if let Some(tx) = self
+            .mempool_handle
+            .call_async(move |m| Box::pin(async move { m.transaction(&id).await }))
+            .await??
+        {
+            self.messaging_handle
+                .send_message(self.id(), SyncMessage::TransactionResponse(tx))?;
+        }
+
+        Ok(())
+    }
+
+    async fn handle_transaction_response(&mut self, tx: SignedTransaction) -> Result<()> {
+        let id = tx.serialized_hash();
+
+        if self.announced_transactions.take(&id).is_none() {
```

Comment:
Isn't it better to implement this as a push/pop queue instead of using this unnecessary invariant?

Reply:
I'm not sure. What if we receive responses in a different order or don't receive one of them at all? With the current implementation we will still be able to obtain other transactions.

Reply:
Receiving responses in a different order should have zero bearing on the final outcome. We cannot guarantee that anyway due to both multithreading and network latency.